src/utils/translator.rs

```diff
@@ -0,0 +1,494 @@
+        (map $arg:ident table_index) => (t.remap(Item::Table, *$arg)?);
+        (map $arg:ident table) => (t.remap(Item::Table, *$arg)?);
+        (map $arg:ident dst_table) => (t.remap(Item::Table, *$arg)?);
+        (map $arg:ident src_table) => (t.remap(Item::Table, *$arg)?);
+        (map $arg:ident type_index) => (t.remap(Item::Type, *$arg)?);
+        (map $arg:ident global_index) => (t.remap(Item::Global, *$arg)?);
+        (map $arg:ident mem) => (t.remap(Item::Memory, *$arg)?);
+        (map $arg:ident src_mem) => (t.remap(Item::Memory, *$arg)?);
+        (map $arg:ident dst_mem) => (t.remap(Item::Memory, *$arg)?);
+        (map $arg:ident data_index) => (t.remap(Item::Data, *$arg)?);
+        (map $arg:ident elem_index) => (t.remap(Item::Element, *$arg)?);
+        (map $arg:ident blockty) => (t.translate_block_type($arg)?);
+        (map $arg:ident relative_depth) => (*$arg);
+        (map $arg:ident targets) => ((
+            $arg
+                .targets()
+                .collect::<wasmparser::Result<Vec<_>, wasmparser::BinaryReaderError>>()?
+                .into(),
+            $arg.default(),
+        ));
+        (map $arg:ident table_byte) => (());
+        (map $arg:ident mem_byte) => (());
+        (map $arg:ident flags) => (());
+        (map $arg:ident ty) => (t.translate_ty($arg)?);
+        (map $arg:ident hty) => (t.translate_heapty($arg)?);
+        (map $arg:ident memarg) => (t.translate_memarg($arg)?);
+        (map $arg:ident local_index) => (*$arg);
+        (map $arg:ident value) => ($arg);
+        (map $arg:ident lane) => (*$arg);
+        (map $arg:ident lanes) => (*$arg);
+
+        // This case takes the arguments of a wasmparser instruction and creates
+        // a wasm-encoder instruction. There are a few special cases for where
+        // the structure of a wasmparser instruction differs from that of
+        // wasm-encoder.
+        (build $op:ident) => (I::$op);
+        (build BrTable $arg:ident) => (I::BrTable($arg.0, $arg.1));
+        (build I32Const $arg:ident) => (I::I32Const(*$arg));
+        (build I64Const $arg:ident) => (I::I64Const(*$arg));
+        (build F32Const $arg:ident) => (I::F32Const(f32::from_bits($arg.bits())));
+        (build F64Const $arg:ident) => (I::F64Const(f64::from_bits($arg.bits())));
+        (build V128Const $arg:ident) => (I::V128Const($arg.i128()));
+        (build $op:ident $arg:ident) => (I::$op($arg));
+        (build CallIndirect $ty:ident $table:ident $_:ident) => (I::CallIndirect {
+            ty: $ty,
+            table: $table,
+        });
+        (build ReturnCallIndirect $ty:ident $table:ident) => (I::ReturnCallIndirect {
+            ty: $ty,
+            table: $table,
+        });
+        (build MemoryGrow $mem:ident $_:ident) => (I::MemoryGrow($mem));
+        (build MemorySize $mem:ident $_:ident) => (I::MemorySize($mem));
+        (build $op:ident $($arg:ident)*) => (I::$op { $($arg),* });
+    }
+
+	wasmparser::for_each_operator!(translate)
```

Comment:
To some extend, I prefer the original version. You have an explicit, raw mapping for each of the wasmparser operator. This highly convoluted, text-based logic seems smart but in fact fragile.

Reply:
It simply tries to reuse the convenience macro [for_each_operator](https://docs.rs/wasmparser/latest/wasmparser/macro.for_each_operator.html).
But agree, this part is not clear. Will restore original one.